client/app/visualizations/table/index.js

```diff
@@ -135,99 +138,75 @@ function getColumnsToDisplay(columns, options, clientConfig) {
   return _.sortBy(_.filter(result, 'visible'), 'order');
 }
 
-function GridRenderer(clientConfig) {
-  return {
-    restrict: 'E',
-    scope: {
-      queryResult: '=',
-      options: '=',
-    },
-    template,
-    replace: false,
-    controller($scope) {
-      $scope.gridColumns = [];
-      $scope.gridRows = [];
-
-      function update() {
-        if ($scope.queryResult.getData() == null) {
-          $scope.gridColumns = [];
-          $scope.filters = [];
-        } else {
-          $scope.filters = $scope.queryResult.getFilters();
-          $scope.gridRows = $scope.queryResult.getData();
-          const columns = $scope.queryResult.getColumns();
-          const columnsOptions = getColumnsOptions(columns, _.extend({}, $scope.options).columns);
-          $scope.gridColumns = getColumnsToDisplay(columns, columnsOptions, clientConfig);
-        }
+const GridRenderer = {
+  bindings: {
+    data: '<',
+    options: '<',
+  },
+  template,
+  controller($scope) {
+    const update = () => {
+      this.gridColumns = [];
+      this.gridRows = [];
+      if (this.data) {
+        this.gridColumns = getColumnsToDisplay(this.data.columns, this.options.columns);
+        this.gridRows = this.data.rows;
       }
+    };
+    update();
 
-      $scope.$watch('queryResult && queryResult.getData()', (queryResult) => {
-        if (queryResult) {
-          update();
-        }
-      });
-
-      $scope.$watch('options', (newValue, oldValue) => {
-        if (newValue !== oldValue) {
-          update();
-        }
-      }, true);
-    },
-  };
-}
+    $scope.$watch('$ctrl.data', update);
```

Comment:
Now that this is a component, you should use `$onChanges` instead of `$watch`.

Reply:
I need to watch `options` for deep changes, so `$onChanges` is not suitable (or will require more work to do); the same in editor. I think, very soon we'll migrate everything to React, so it's okay to use `$scope.$watch` for now